src/entities/createActionListenerMiddleware.ts

```diff
@@ -0,0 +1,268 @@
+import { Middleware, Dispatch, AnyAction, MiddlewareAPI, Action } from 'redux'
+import { TypedActionCreator } from '../mapBuilders'
+import { createAction, BaseActionCreator } from '../createAction'
+
+type ActionListener<A extends AnyAction, S, D extends Dispatch<AnyAction>> = (
+  action: A,
+  api: MiddlewareAPI<D, S>
+) => void
+interface ActionListenerOptions<
+  A extends AnyAction,
+  S,
+  _ extends Dispatch<AnyAction>
+> {
+  /**
+   * Indicates that the listener should be removed after if has run once.
+   */
+  once?: boolean
+  /**
+   * If set to true, the action will not be forwarded to
+   * * listeners that were registered after this listener
+   * * middlewares later in the middleware chain
+   * * reducers
+   * If this listener is skipped due to `options.condition`, this has no effect.
+   */
+  preventPropagation?: boolean
+  /**
+   * A function that determines if the listener should run, depending on the action and probably the state.
+   */
+  condition?(action: A, getState: () => S): boolean
+}
+
+interface AddListenerAction<
+  A extends AnyAction,
+  S,
+  D extends Dispatch<AnyAction>
+> {
+  type: 'actionListenerMiddleware/add'
+  payload: {
+    type: string
+  }
+  meta: {
+    listener: ActionListener<A, S, D>
+    options?: ActionListenerOptions<A, S, D>
+  }
+}
+
+export const addListenerAction = createAction(
+  'actionListenerMiddleware/add',
+  function prepare(
+    typeOrActionCreator: string | TypedActionCreator<string>,
+    listener: ActionListener<any, any, any>,
+    options?: ActionListenerOptions<AnyAction, any, any>
+  ) {
+    const type =
+      typeof typeOrActionCreator === 'string'
+        ? typeOrActionCreator
+        : (typeOrActionCreator as TypedActionCreator<string>).type
+
+    return {
+      payload: {
+        type
+      },
+      meta: {
```

Comment:
Feels like `listener` and `options` would be better suited for `payload`, conceptually. After all, those are key aspects of this particular action, and the middleware is explicitly using them.

Reply:
Yeah, no particular reason to put these in `meta` tbh, other than that I personally try to keep non-serializable stuff out of `payload`.

Reply:
:heavy_check_mark: